From 0e34eb614092f74f87a6260465eee455afdb1fc6 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:38:49 -0400
Subject: [ethernet] i40e: Delete stale MAC filters after change

Message-id: <1409658042-22689-169-git-send-email-sassmann@redhat.com>
Patchwork-id: 90362
O-Subject: [RHEL7.1 PATCH 168/281] i40e: Delete stale MAC filters after change
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Greg Rose <gregory.v.rose@intel.com>

Delete all the old and stale MAC filters for the VF VSI when the host
administrator changes the VF MAC address from under its feet.  Also don't
bother to add a filter for the VSI when its going to go away anyway.
Just record the new address and punch the VF reset.

Change-ID: Ic0d12055926f41989d1965ccf500053729c063ad
Signed-off-by: Greg Rose <gregory.v.rose@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit 29f71bb0915db39c493235b63abf6b03150db84f)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_virtchnl_pf.c b/drivers/net/ethernet/intel/i40e/i40e_virtchnl_pf.c
index 26fd221..25280d4 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_virtchnl_pf.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_virtchnl_pf.c
@@ -2063,14 +2063,11 @@ int i40e_ndo_set_vf_mac(struct net_device *netdev, int vf_id, u8 *mac)
  i40e_del_filter(vsi, vf->default_lan_addr.addr, vf->port_vlan_id,
    true, false);
 
- /* add the new mac address */
- f = i40e_add_filter(vsi, mac, vf->port_vlan_id, true, false);
- if (!f) {
-  dev_err(&pf->pdev->dev,
-   "Unable to add VF ucast filter\n");
-  ret = -ENOMEM;
-  goto error_param;
- }
+ /* Delete all the filters for this VSI - we're going to kill it
+  * anyway.
+  */
+ list_for_each_entry(f, &vsi->mac_filter_list, list)
+  i40e_del_filter(vsi, f->macaddr, f->vlan, true, false);
 
  dev_info(&pf->pdev->dev, "Setting MAC %pM on VF %d\n", mac, vf_id);
  /* program mac filter */
-- 
1.7.1